lib/rust/lexer/definition/src/lexer.rs

```diff
@@ -464,9 +722,805 @@ impl EnsoLexer {
 #[allow(dead_code)]
 impl EnsoLexer {
 
+    // === Error ===
+
+    /// Triggered when encountering an invalid set of quotes.
+    fn on_invalid_quote<R:LazyReader>(&mut self, _reader:&mut R) {
+        trace!(self.logger,"Text::on_invalid_quote");
+        let offset     = self.offset.consume();
+        let bad_quotes = self.consume_current();
+        let token      = Token::InvalidQuote(bad_quotes,offset);
+        self.append_token(token);
+    }
+
+    /// Submit a missing quote in a nested text line literal.
+    fn submit_missing_quote_nested<R:LazyReader>
+    ( &mut self
+      , line_ending:token::LineEnding
+      , reader:&mut R
+    ) {
+        trace!(self.logger,"Text::submit_missing_quote_nested");
+        let interpolate_is_closed = false;
+        let matched_bookmark      = self.bookmarks.matched_bookmark;
+        self.bookmarks.rewind(matched_bookmark,reader);
+        let nested         = self.text_state.unsafe_current_mut();
+        nested.style       = Some(token::TextStyle::UnclosedLine);
+        let nested_text    = self.text_state.unsafe_end_literal();
+        let nested_literal = nested_text.into();
+        self.append_token(nested_literal);
+        self.pop_state();
+        self.on_interpolate_end(interpolate_is_closed,reader);
+        let top_level       = self.text_state.unsafe_current_mut();
+        let top_level_style = top_level.style.expect("Must be set.");
+        top_level.style = match top_level_style {
+            token::TextStyle::FormatLine => Some(token::TextStyle::UnclosedLine),
+            _                            => Some(top_level_style),
+        };
+        let text    = self.text_state.unsafe_end_literal();
+        let literal = text.into();
+        self.append_token(literal);
+        self.pop_state();
+        match line_ending {
+            token::LineEnding::None => self.on_eof(reader),
+            _                       => self.block_state.push_line_ending(line_ending),
+        }
+    }
+
+    /// Submit a missing closing quote in a text line literal.
+    fn submit_missing_quote<R:LazyReader>(&mut self, new_line:token::LineEnding, reader:&mut R) {
+        trace!(self.logger,"Text::submit_missing_quote");
+        let matched_bookmark = self.bookmarks.matched_bookmark;
+        self.bookmarks.rewind(matched_bookmark, reader);
+        let current   = self.text_state.unsafe_current_mut();
+        current.style = Some(token::TextStyle::UnclosedLine);
+        let text      = self.text_state.unsafe_end_literal();
+        let literal   = text.into();
+        self.append_token(literal);
+        self.pop_state();
+        match new_line {
+            token::LineEnding::None => self.on_eof(reader),
+            _                       => self.block_state.push_line_ending(new_line),
+        }
```

Comment:
this function and one above look very similar - please refactor out the common part